docs/settings.rst
Settings
========

Basic settings
--------------

.. _PYBB_TOPIC_PAGE_SIZE:

PYBB_TOPIC_PAGE_SIZE
....................

Number of posts in topic page

Default: 10

.. _PYBB_FORUM_PAGE_SIZE:

PYBB_FORUM_PAGE_SIZE
....................

Number of topics in forum page

Default: 10

.. _PYBB_FREEZE_FIRST_POST:

PYBB_FREEZE_FIRST_POST
......................

Freeze first post in topics (show on every page)

Default: False

.. _PYBB_DEFAULT_TITLE:

PYBB_DEFAULT_TITLE
..................

Default title for forum index page

Default: 'PYBB Powered Forum'

.. _PYBB_DEFAULT_AUTOSUBSCRIBE:

PYBB_DEFAULT_AUTOSUBSCRIBE
..........................

Users will be automatically subscribed to topic when create post in it.

Default: True

.. _PYBB_DISABLE_SUBSCRIPTIONS:

PYBB_DISABLE_SUBSCRIPTIONS
..........................

Users won't be able to subscribe to topic.
If you want to have a more advanced mode than enable / disable (for example, use model permissions),
you just have to overwrite the "may_subscribe_topic" method of the Permission handler.
If you disabled topic subscriptions, already subscribed users will still receive notifications:
see ``PYBB_DISABLE_NOTIFICATIONS`` to stop notifications sending.

Default: False

.. _PYBB_DISABLE_NOTIFICATIONS:

PYBB_DISABLE_NOTIFICATIONS
..........................

Users which have subscribed to a topic won't receive notifications but still be able to subscribe
to topics. See ``PYBB_DISABLE_NOTIFICATIONS`` to disable topic subscription too.
This is usefull if you want to to use your own notification system. Additionally, if your custom
user model has a ``receive_emails`` field, it will be used to determine whether to send notification
emails to each user.

Default: False

.. _PYBB_USE_DJANGO_MAILER:

PYBB_USE_DJANGO_MAILER
......................

When True and django-mailer app installed, then for sending email pybbm will use this app. With django-mailer you can
manage emails from your site in queue. But in this case you have to periodically actually
send emails from queue. For more information see `app home page <https://github.com/pinax/django-mailer/>`_.

Default: False


Emoticons
---------

.. _PYBB_SMILES_PREFIX:

PYBB_SMILES_PREFIX
..................

Prefix for emoticons images set, related to STATIC_URL.

Default: 'pybb/emoticons'

.. _PYBB_SMILES:

PYBB_SMILES
...........

Dict for emoticon replacement.
Key - text to be replaced, value - image name.

Default::

    {
        '&gt;_&lt;': 'angry.png',
        ':.(': 'cry.png',
        'o_O': 'eyes.png',
        '[]_[]': 'geek.png',
        '8)': 'glasses.png',
        ':D': 'lol.png',
        ':(': 'sad.png',
        ':O': 'shok.png',
        '-_-': 'shy.png',
        ':)': 'smile.png',
        ':P': 'tongue.png',
        ';)': 'wink.png'
    }

e.g. text  ";)" in post will be replaced to::

    <img src="{{ STATIC_URL }}{{ PYBB_SMILES_PREFIX }}wink.png">

with default setting.


User profile settings
---------------------

Next settings used only if you don't customize user profile model,
user profile creation form or templates.

.. _PYBB_PROFILE_RELATED_NAME:

PYBB_PROFILE_RELATED_NAME
.........................

Related name from profile's OneToOne relationship to User model. If profile model is User
model itself then set it to `None`.

Default: 'pybb_profile'

For more information see :doc:`how to use custom user model with pybbm</customuser>`

.. _PYBB_AVATAR_WIDTH:

PYBB_AVATAR_WIDTH
.................

Avatar width to use in templates (avatars scaled using sorl.thumbnail
if it installed and included in project).

Default: 80

.. _PYBB_AVATAR_HEIGHT:

PYBB_AVATAR_HEIGHT
..................

Avatar height to use in templates (avatars scaled using sorl.thumbnail
if it installed and included in project)

Default: 80

.. _PYBB_MAX_AVATAR_SIZE:

PYBB_MAX_AVATAR_SIZE
....................

Maximum avatar size, in bytes

Default: 51200 (50 Kb)

.. _PYBB_DEFAULT_TIME_ZONE:

PYBB_DEFAULT_TIME_ZONE
......................

Default time zone for forum as integer. E.g. setting to 1 means GMT+1 zone.

Default: 3 (Moscow)

.. _PYBB_SIGNATURE_MAX_LENGTH:

PYBB_SIGNATURE_MAX_LENGTH
.........................

Limit of symbols in user signature

Default: 1024

.. _PYBB_SIGNATURE_MAX_LINES:

PYBB_SIGNATURE_MAX_LINES
........................

Limit of lines in user signature

Default: 3

.. _PYBB_DEFAULT_AVATAR_URL:

PYBB_DEFAULT_AVATAR_URL
.......................

Will be used if user doesn't upload avatar

Default: settings.STATIC_URL + 'pybb/img/default_avatar.jpg'

Style
-----

You can use builtin templates with custom basic template.

.. _PYBB_TEMPLATE:

PYBB_TEMPLATE
.............

Builtin templates will inherit this template

Default: 'base.html

.. _PYBB_TEMPLATE_MAIL_TXT:

PYBB_TEMPLATE_MAIL_TXT
.............

Builtin `txt` emails templates will inherit this template

Default: 'pybb/mail_templates/base.html

.. _PYBB_TEMPLATE_MAIL_HTML:

PYBB_TEMPLATE_MAIL_HTML
.............

Builtin `html` emails templates will inherit this template

Default: 'pybb/mail_templates/base-html.html


Markup engines
--------------

.. _PYBB_MARKUP:

PYBB_MARKUP
...........

Markup engine used in forum. Also see :ref:`PYBB_MARKUP_ENGINES` below

Default: 'bbcode`

.. _PYBB_MARKUP_ENGINES_PATHS:

PYBB_MARKUP_ENGINES_PATHS
.........................

Dict with avaiable markup engines path. One of them should be selected with PYBB_MARKUP

Markup engine should be a path to a class, that inherits from `pybb.markup.base.BaseParser`.
Markup engine should take care of replacing smiles in body with related emoticons.

by default PyBBM support `bbcode` and `markdown` markup::

    {
        'bbcode': 'pybb.markup.bbcode.BBCodeParser',
        'markdown': 'pybb.markup.markdown.MarkdownParser'
    }

Please note, that previous version of pybb used two different settings : 
`PYBB_MARKUP_ENGINES` and `PYBB_QUOTE_ENGINES` which were callables.
This is still supported, but is deprecated.

.. _PYBB_MARKUP_ENGINES:

PYBB_MARKUP_ENGINES (deprecated)
................................

Should be the same dict with paths to markup engine classes as `PYBB_MARKUP_ENGINES_PATH` setting

Default: `PYBB_MARKUP_ENGINES_PATHS`.

For more information see :doc:`markup`

.. _PYBB_QUOTE_ENGINES:

PYBB_QUOTE_ENGINES (deprecated)
...............................

**Deprecation note: Every markup class must inherit from** `pybb.markup.base.BaseParser`

**For more information see** :doc:`markup`

Should be the same dict with paths to markup engine classes as `PYBB_MARKUP_ENGINES_PATH` setting

Default: `PYBB_MARKUP_ENGINES_PATHS`.


Post cleaning/validation
------------------------

.. _PYBB_BODY_CLEANERS:

PYBB_BODY_CLEANERS
..................

List of paths to 'cleaner' functions for body post to automatically remove undesirable content from posts.
Cleaners are user-aware, so you can disable them for some types of users.

Each function in list should accept `auth.User` instance as first argument and `string` instance as second, returned value will be sended to next function on list or saved and rendered as post body.

For example this is enabled by default `rstrip_str` cleaner::

    def rstrip_str(user, str):
        if user.is_staff:
            return str
        return '\n'.join([s.rstrip() for s in str.splitlines()])

Default::

    PYBB_BODY_CLEANERS = ['pybb.markup.base.rstrip_str', 'pybb.markup.base.filter_blanks']

.. _PYBB_BODY_VALIDATOR:

PYBB_BODY_VALIDATOR
...................

Extra form validation for body of post.

Called as::

    PYBB_BODY_VALIDATOR(user, body)

at `clean_body` method of `PostForm` Here you can do various checks based on user stats.
E.g. allow moderators to post links and don't allow others. By raising::

    forms.ValidationError('Here Error Message')

You can show user what is going wrong during validation.

You can use it for example for time limit between posts, preventing URLs, etc.

Default: None


Anonymous/guest posting
-----------------------

.. _PYBB_ENABLE_ANONYMOUS_POST:

PYBB_ENABLE_ANONYMOUS_POST
..........................

Allow post for not-authenticated users.

Default: False

See :doc:`anonymous posting</anonymous>` for details.

.. _PYBB_ANONYMOUS_USERNAME:

PYBB_ANONYMOUS_USERNAME
.......................

Username for anonymous posts. If no user with this username exists it will be created on first anonymous post.

Default: 'Anonymous'

.. _PYBB_ANONYMOUS_VIEWS_CACHE_BUFFER:

PYBB_ANONYMOUS_VIEWS_CACHE_BUFFER
.................................

Number of anonymous views for each topic, that will be cached. For disabling caching anonymous views
just set it to `None`.

Default: 100


Premoderation
-------------

.. _PYBB_PREMODERATION:

PYBB_PREMODERATION
..................

Filter for messages that require pre-moderation.

Default: False

See :doc:`Pre-moderation</premoderation>` for details.


Attachments
-----------

.. _PYBB_ATTACHMENT_ENABLE:

PYBB_ATTACHMENT_ENABLE
......................

Enable attahcments for all users.

Default: False

.. _PYBB_ATTACHMENT_SIZE_LIMIT:

PYBB_ATTACHMENT_SIZE_LIMIT
..........................

Maximum attachment limit (in bytes).

Default: 1048576 (1Mb)

.. _PYBB_ATTACHMENT_UPLOAD_TO:

PYBB_ATTACHMENT_UPLOAD_TO
.........................

Directory in your media path for uploaded attacments.

Default: 'pybb_upload/attachments'

Polls
-----

Note: For disabling polls on your forum, write custom permission handler and return from `may_create_poll` method `False`
See `PYBB_PERMISSION_HANDLER` setting.

.. _PYBB_POLL_MAX_ANSWERS:

PYBB_POLL_MAX_ANSWERS
.....................

Max count of answers, that user can add to topic.

Default: 10


Permissions
-----------

.. _PYBB_AUTO_USER_PERMISSIONS:

PYBB_AUTO_USER_PERMISSIONS
..........................

Automatically adds add post and add topic permissions to users on user.save().

Default: True

.. _PYBB_PERMISSION_HANDLER:

PYBB_PERMISSION_HANDLER
.......................

If you need custom permissions (for example, private forums based on application-specific 
user groups), you can set :ref:`PYBB_PERMISSION_HANDLER` to a class which inherits from
`pybb.permissions.DefaultPermissionHandler` (default), and override any of the `filter_*` and
`may_*` method. For details, look at the source of `pybb.permissions.DefaultPermissionHandler`.
All methods from permission handler (custom or default) can be used in templates as filters,
if loaded pybb_tags. In template will be loaded methods which start with 'may' or 'filter'
and with three or two arguments (include 'self' argument)

Default: 'pybb.permissions.DefaultPermissionHandler'


Urls
----

.. _PYBB_NICE_URL:

PYBB_NICE_URL
.............

Changes old/classics URLs to more semantic URLs using Category/Forum/Topic's slug.
For example www.yourforum.com/forum/1 becomes www.yourforum.com/c/category_slug/forum_slug.
Old URLs will have a permanent redirections to new ones.

Default: False

.. _PYBB_NICE_URL_PERMANENT_REDIRECT:

PYBB_NICE_URL_PERMANENT_REDIRECT
................................

When PYBB is set to use PYBB_NICE_URL, this setting changes the HTTP response code used
to redirect old style URL to new one. True (default) use 301 (permantent) redirect. If set to False,
it uses 302 (temporary) redirect. `False` value is usefull for testing period to not loose SEO related
to old URLs, then, once testing period is over, setting this to True will ensure that your old URLs will
be updated to the new ones next time the Search Engine will check it.

Default: True

.. _PYBB_NICE_URL_SLUG_DUPLICATE_LIMIT:

PYBB_NICE_URL_SLUG_DUPLICATE_LIMIT
..................................

Limit for checking duplicate slugs.
After reaching this limit while trying to find unique slug ValidationError will be raised.

Default: 100

.. PYBB_ENABLE_ADMIN_POST_FORM:

PYBB_ENABLE_ADMIN_POST_FORM
...........................

Enable admin post form that allowed staff to post with any username automagically creating it if it did not exist.

Default: True

.. PYBB_ALLOW_DELETE_OWN_POST:

PYBB_ALLOW_DELETE_OWN_POST
..........................

Allow non-superusers to delete their own posts.

Default: True


CHANGELOG
0.19.0
    * PyBBM is now compatible with Django 1.8, 1.11, 2.0
    * Fix permissions and add documentation for default permissions
    * Allow moderators to move a topic or split posts
    * Update translations
    * Multiple other fixes, refactoring
0.18.4
    * Fix misspelling in Swedish translation
0.18.3
    * Fix 0005 and 0006 migrations (changes will not affect db, so it's ok if this migrations were already executed)
0.18.2
    * Minor fixes.
0.18.1
    * Minor fixes.
0.18
    * PyBBM is now compatible with Django>=1.8,<1.11
    * Allow non-moderators to delete their own posts.
    * Add setting to enable or disable admin post form.
    * Add Swedish translation.
    * Use FileField instead of ImageField when pillow is not available to make pillow depencency optional.
    * Use staticfiles in all templates.
    * Improve permission checking.
    * Allow users to subscribe to a forum.
    * Add a form to grant users moderator privileges.
    * Add notification's emails HTML alternative.
    * Add the ability to the user to use their attachments inside their posts to render it as link, image etc.
    * Multiple fixes and improvements.
0.17.3
    * Fast fix for migrations for Posgres database.
      If you already get and applied migrations from 0.17 version (for example on MySQL DB) you can skip new
      migrations with `manage.py migrate pybb --fake`
0.17
    * Topic and post creation wrapped in transaction
    * All topic/post/poll related forms can be overrided when custom view inherites pybbm view
    * Demo data for example projects
    * Using active markup engine when quoting posts via javascript
    * Functionality to support disabling default pybbm subscriptions and notifications and
      new settings: `PYBB_DISABLE_SUBSCRIPTIONS` and `PYBB_DISABLE_NOTIFICATIONS`
    * Fixed sorl.thumbnail/easy_thumbnail compatibility in standard `pybb/avatar.html` template
    * Improved example projects
    * Removed applying `urlize` filter over html produced by markdown parser
      (it doesn't play nicely with html markup as noted in django's docs)
    * django 1.8 compatibility
    * common django layout for test project
    * use mysqlclient package for testing installation with mysql database backend on python 3
    * optional enabling "nice urls" for entire forum,
      that looks like '<forum prefix>/c/<category slug>/<forum slug>/<topic slug>/'
0.16.1
    * Fast fixes
0.16
    * Django 1.7 compatibility.
    * Fixed creating custom profile model of any class defined in settings with right related name to user model.
      *Migration note*: If you have workaround for creating profile in your code, you should remove it for
      preventing possible dubplicate unique key error on user creating.
    * New get_display_name method for profile model used to unification displaying username through forum
    * New markup processing. See `markup`
0.15.6
    * Make all migrations compatible with custom user model. Break dependency on sorl.thumbnail in migrations
    * Compatibility functions moved to compat.py module
    * Email notifications optimization
    * Example_bootstrap projects now based on bootstrap 3
    * Fixes and improvements
0.15.5
    * Fixed bug when user can vote (or cancel vote) when topic was closed.
    * Added `may_vote_in_topic` method to permission handler.
    * Fixed blocking user view
0.15.4
    * Hot fixes to bbcode transform
0.15.3
    * bbcode engine simplified
0.15.2
    * Pybbm specific forms moved to views' attributes, added new functions to views to get such forms dynamically.
      This makes overriding pybbm forms much easier
    * Moving from unmaintained postmarkup package to bbcode project as default bbcode render engine
      Changed output html for [code] tag. It will be <code></code> tags instead of <div class="code"></div>.
      So you should duplicate styles applied to div.code for code html tag.
    * Japanese translation
0.15.1
    * Hot fixes for Python 3 support
    * Fixes for Chinese translation
0.15
    * Python 3 support
    * Chinese translation
0.14.9
    * Two new methods added to permission handler: `may_attach_files` and `may_create_poll`. First method used for
      restrict attaching files to post by user. By default it depends on `PYBB_ATTACHMENT_ENABLE` setting.
      Second may be used to restrict some users to create/edit polls. By default it always return `True`.
      For disabling polls on your forum, just write custom permission handler and return from this method `False`
0.14.8
    * Improved javascript functionality: quote selected text, qoute full original message via ajax,
      insert nickname in post body. For enabling this functionality you should satisfy `some requirements</javascript>`
      in your templates
    * Support for nested forums
    * `PybbProfile` abstract model moved to `pybb.profiles` module to avoid circular imports when checking models.
0.14.7
    * Django 1.6 compatibility
    * unblock user functionality added
0.14.6
    * Cache anonymous views count for topic and save it in database only when some count reached (100 by default).
      This value can be changed by setting `PYBB_ANONYMOUS_VIEWS_CACHE_BUFFER`. Also added custom filter
      `pybbm_calc_topic_views` that calc actual views count for topic
    * Fix for migration that may fails on clean mysql installation
    * Fixed perfomance issue with feed views
    * Using custom permissions handler in feed views
0.14.5
    * Minor fixes
0.14.4
    * Fix for migration that may fails on clean mysql installation (not fixed really, filxed after 0.14.5)
    * Make example_thirdparty project bootstrap3 compatible
0.14.3
    * Show only available topics (by permission handler) in ForumView
0.14.2
    * Fixed MultipleObjectReturned when topic has more than one moderator
0.14.1
    * Fixed circular import issue
0.14
    * Restored views for rendering user's posts and topics and link to that views from profile info page

pybb/settings.py
from os.path import join
from annoying.functions import get_config

PYBB_TOPIC_PAGE_SIZE = get_config('PYBB_TOPIC_PAGE_SIZE', 10)
PYBB_FORUM_PAGE_SIZE = get_config('PYBB_FORUM_PAGE_SIZE', 10)
PYBB_USERS_PAGE_SIZE = get_config('PYBB_USERS_PAGE_SZIE', 20)
PYBB_AVATAR_WIDTH = get_config('PYBB_AVATAR_WIDTH', 60)
PYBB_AVATAR_HEIGHT = get_config('PYBB_AVATAR_HEIGHT',60)
PYBB_MAX_AVATAR_SIZE = get_config('PYBB_MAX_AVATAR_SIZE', 1024*50)
PYBB_DEFAULT_TIME_ZONE = get_config('PYBB_DEFAULT_TIME_ZONE', 3)

PYBB_SIGNATURE_MAX_LENGTH = get_config('PYBB_SIGNATURE_MAX_LENGTH', 1024)
PYBB_SIGNATURE_MAX_LINES = get_config('PYBB_SIGNATURE_MAX_LINES', 3)

PYBB_READ_TIMEOUT = get_config('PYBB_READ_TIMEOUT', 3600 * 24 * 7) # seconds

PYBB_DEFAULT_MARKUP = get_config('PYBB_DEFAULT_MARKUP', 'bbcode')
PYBB_FREEZE_FIRST_POST = get_config('PYBB_FREEZE_FIRST_POST', False)

PYBB_ATTACHMENT_SIZE_LIMIT = get_config('PYBB_ATTACHMENT_SIZE_LIMIT', 1024 * 1024)
PYBB_ATTACHMENT_ENABLE = get_config('PYBB_ATTACHMENT_ENABLE', False)
PYBB_ATTACHMENT_UPLOAD_TO = get_config('PYBB_ATTACHMENT_UPLOAD_TO', join('pybb_upload', 'attachments'))

PYBB_DEFAULT_AVATAR_URL = get_config('MEDIA_URL', '') + 'pybb/img/default_avatar.jpg'

PYBB_DEFAULT_TITLE = get_config('PYBB_DEFAULT_TITLE', 'PYBB Powered Forum')

from postmarkup import render_bbcode
from markdown import Markdown
from django.utils.html import urlize

PYBB_SMILES_PREFIX = 'pybb/emoticons/'

PYBB_SMILES = get_config('PYBB_SMILES', {
    '&gt;_&lt;': 'angry.png',
    ':.(': 'cry.png',
    'o_O': 'eyes.png',
    '[]_[]': 'geek.png',
    '8)': 'glasses.png',
    ':D': 'lol.png',
    ':(': 'sad.png',
    ':O': 'shok.png',
    '-_-': 'shy.png',
    ':)': 'smile.png',
    ':P': 'tongue.png',
    ';)': 'wink.png'
})

MEDIA_URL = get_config('MEDIA_URL', '/media/')

def smile_it(str):
    s = str
    for smile, url in PYBB_SMILES.items():
        s = s.replace(smile, '<img src="%s%s%s">' % (MEDIA_URL, PYBB_SMILES_PREFIX, url))
    return s

PYBB_MARKUP_ENGINES = get_config('PYBB_MARKUP_ENGINES', {
    'bbcode': lambda str: urlize(smile_it(render_bbcode(str))),
    'markdown': lambda str: urlize(smile_it(Markdown(safe_mode='escape').convert(str)))
})

PYBB_QUOTE_ENGINES = get_config('PYBB_QUOTE_ENGINES', {
    'bbcode': lambda text, username="": '[quote="%s"]%s[/quote]\n' % (username, text),
    'markdown': lambda text, username="": '>'+text.replace('\n','\n>').replace('\r','\n>') + '\n'
})